### ISSUE TITLE
generates empty list error

### ISSUE THREAD

**Issue by vinayakumarr** [renan-campos/MSA#4]
Due to memory problem i have used only 1000 files in train- pos and neg and when i execute python src/train.py, after some time it generate te following eror. How to solve this. Please see the attached file
![untitled](https://cloud.githubusercontent.com/assets/18217467/15803341/631277d2-2af6-11e6-8c21-94f0531bcf63.png)

**Comment by rcampos2029** [comment]
In the original code, we performed 40 batches with 600 files in each (300 pos/300 neg). So your code fails on batch 2 because it runs out of files to train on. I just updated the code to have variables to tune the development set size, number of batches, and batch size on line 46. Change these variables so that DEV_SIZE + BATCH_SIZE*BATCHES = len(train)

**Comment by vinayakumarr** [comment]
500+(50*10)=len(train). Successfully works. test data set is 250. for all test data set it is classifying as 0. How to see the fscore and precision and accuracy. I came to know in tmp folder batch files will be create and that has these information. But all files is showing accuracy precision and recall as 1. Do you where the problem has occurred or whether the results are correct.

**Comment by rcampos2029** [comment]
that is odd that they all show up as 1. What does it say for TP, TN, FP, and FN?  predict.py has the code for the metrics.

**Comment by vinayakumarr** [comment]
FP.txt and FN.txt files doesn't contain anything. But TP.txt and TN.txt contains 1001 lines (includes the name of files). You can see a part of TP.txt and TN.txt in the below attached image

TP
![tn](https://cloud.githubusercontent.com/assets/18217467/15811624/81192e1a-2bc8-11e6-9a65-9b5bc94d1f7c.png)

TN
![tp](https://cloud.githubusercontent.com/assets/18217467/15811665/b4ffe674-2bc8-11e6-8afe-0388ac17aa4d.png)

**Comment by vinayakumarr** [comment]
I think the below code fails to work

precision = len(TP) / float(len(TP)+len(FP))
  recall    = len(TP) / float(len(TP)+len(FN))
  f1        = (2*precision * recall) / (precision + recall)
  accuracy  = (len(TP) + len(TN)) / float(len(TP) + len(TN) + len(FP) + len(FN))
  print >>out, "*** Results ***"
  print >>out, "\tTotal test files: %d" % (len(test_set))
  print >>out, "\t(%d positive, %d negative)" % (len(dev['pos']), len(dev['neg']))
  print >>out, " "
  print >>out, "\tTP: %d TN: %d FP: %d FN: %d" % (len(TP), len(TN), len(FP), len(FN)) 
  print >>out, " "
  print >>out, "\tPrecision: %0.4f" % (precision)
  print >>out, "\tRecall: %0.4f" % (recall)
  print >>out, "\tF1: %0.4f" % (f1)
  print >>out, "\tAccuracy: %0.4f" % (accuracy)

**Comment by vinayakumarr** [comment]
predict.py shows below result
![predict](https://cloud.githubusercontent.com/assets/18217467/15811876/c3c18dae-2bcb-11e6-9015-2d928aec94d0.png)

pos (train) contains 1002 files and neg (train) contains 1002 and test contains 252 files

**Comment by rcampos2029** [comment]
Are the TN and TP pictures right? TP should display positive, and TN should show negative. It's also odd that it lists 2002 files instead of 2004 if there are 1002 files in each directory. Are there any batch_#.txt files in your tmp directory (such as batch_0.txt?

**Comment by vinayakumarr** [comment]
i mentioned wrong in previous post, such that pos (train) contains 1001 files and neg (train) contains 1001 and test contains 251 files

inside src/tmp folder batch_*.txt contains 50 files 

each file has precision, recall, f1 and accuracy. it show total test files as 1000 but test has only 251 files.

what these batch files are? 

How to correct this?

**Comment by rcampos2029** [comment]
So the batch files are calculating the metrics on the development set during each batch iteration.
The reason it says 1000 files is because your dev set was specified as 500, and there was a bug that took 500 from positive and negative instead of 500 total.  
I just updated the code to fix this, so now if you specify a dev size of 500, it will actually be 500. 

Running predict by itself will do the calculations on the entire training set. To classify the test set, run test.py. This will generate a csv file of ID,LABEL.

**Comment by rcampos2029** [comment]
Oops, just completed the push now. Try pulling, changing the SIZE variables and running again.

**Comment by vinayakumarr** [comment]
I have 1000 train/pos and 1000/neg, so i used

DEV_SIZE = 1000    # Split evenly between pos/neg
BATCHES = 20
BATCH_SIZE = 50 

is it correct?

**Comment by rcampos2029** [comment]
Yes.

**Comment by vinayakumarr** [comment]
now again i get the previously solved error
![untitled](https://cloud.githubusercontent.com/assets/18217467/15834085/48914102-2c47-11e6-9d03-24cef6381491.png)

**Comment by rcampos2029** [comment]
That's odd, on line 161 of train.py, add the following just to make sure the script see 2000 files:
`print len(training_set) ; exit()`

Could you show batch_0.txt?

**Comment by vinayakumarr** [comment]
shows 1998

**Comment by rcampos2029** [comment]
That would explain the pop from empty list error, and that means train.py only sees 999 files in each directory (pos/neg).

**Comment by vinayakumarr** [comment]
when i change to 

DEV_SIZE = 998 # Split evenly between pos/neg
BATCHES = 20
BATCH_SIZE = 50

it works fine but no improvement in results. result.csv file conain all values 0. Also batch files shows accuracy, precsion and csv all values as 1.

Is this becoz of less amount dataset to training?

**Comment by rcampos2029** [comment]
Could be, having accuracy, precision and recall as 1.0 for all of the files is really peculiar. Is the training data you're using publicly available? I'd like to give it a try.

**Comment by vinayakumarr** [comment]
I am using your data only. I have taken few data from training and testing. I am trying it. I am not using my dataset.

**Comment by vinayakumarr** [comment]
problem solved. I am able to get the exact results. I have one more question that I am using another language data set. Could you please tell where and all we need to change the code in order to make it work and as well as to increase the accuracy.

**Comment by vinayakumarr** [comment]
(?:[a-z][a-z'\-_]+[a-z]) could you please explain this line.

**Comment by rcampos2029** [comment]
That expression captures words that contains apostrophes and dashes.
How did you fix the issue?

**Comment by vinayakumarr** [comment]
I didn't make any change in code. But I was not properly copied your data set to data folder. train folder had files but when i checked it was not containing any text. Your code works fine. I have another question I want to use it for other language so could you please tell where and all i need to change the code. i think only thing i need change is the following regex expression am i correct (?:[a-z][a-z'-_]+[a-z])

**Comment by rcampos2029** [comment]
Yes that is what you would need to change.

**Comment by vinayakumarr** [comment]
http://tamilcube.com/learn-tamil/tamil-alphabets-chart.aspx 
Please have a link the above link shows the alphabets in Tamil language.

(?:[அ-ஔ][அ-ஔ'\-_]+[அ-ஔ]) I just replace like this and it generates the error. Please look at the below attached image for the error. 
![untitled](https://cloud.githubusercontent.com/assets/18217467/15859108/5c5844fc-2ce1-11e6-9e4e-a6a7c48d8260.png)

How to correct this?

Also in tokenizer.py i just added encoding as utf-8 then also it generates an error in data.py. You can look the below attached image for that error.
![untitled](https://cloud.githubusercontent.com/assets/18217467/15859210/c96fbc96-2ce1-11e6-973f-d9670a3b8a73.png)

**Comment by vinayakumarr** [comment]
I think your code will not work for above case. Am I correct

**Comment by rcampos2029** [comment]
I'm not sure, I haven't dealt with different non-ASCII alphabets before. 

The data error shown has to do with the file name, they have to be of the form <ID>_<RANK>.txt or at least <ID>.txt with no non-numeric characters.

**Comment by vinayakumarr** [comment]
k thanks. I will try once i am able to run the code successfully, i will share it with you,
